src/environmentd/src/http/sql.rs

```diff
@@ -7,33 +7,124 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
+use std::time::Duration;
+
 use anyhow::anyhow;
+use async_trait::async_trait;
+use axum::extract::ws::{Message, WebSocket};
+use axum::extract::WebSocketUpgrade;
 use axum::response::IntoResponse;
 use axum::Json;
+use futures::Future;
 use http::StatusCode;
 use itertools::izip;
+use mz_ore::iter::IteratorExt;
 use serde::{Deserialize, Serialize};
 
-use mz_adapter::session::{EndTransactionAction, TransactionStatus};
+use mz_adapter::session::{EndTransactionAction, RowBatchStream, TransactionStatus};
 use mz_adapter::{ExecuteResponse, ExecuteResponseKind, PeekResponseUnary, SessionClient};
+use mz_ore::result::ResultExt;
 use mz_pgwire::Severity;
 use mz_repr::{Datum, RowArena};
 use mz_sql::ast::display::AstDisplay;
 use mz_sql::ast::{Raw, Statement, StatementKind};
 use mz_sql::plan::Plan;
+use tokio::time;
 
 use crate::http::AuthedClient;
 
 pub async fn handle_sql(
     mut client: AuthedClient,
     Json(request): Json<SqlRequest>,
 ) -> impl IntoResponse {
-    match execute_request(&mut client.0, request).await {
-        Ok(res) => Ok(Json(res)),
+    let mut res = SqlResponse {
+        results: Vec::new(),
+    };
+    match execute_request(&mut client, request, &mut res).await {
+        Ok(()) => Ok(Json(res)),
         Err(e) => Err((StatusCode::BAD_REQUEST, e.to_string())),
     }
 }
 
+#[derive(Serialize)]
+struct ErrorResponse {
+    error: String,
+}
+
+pub async fn handle_sql_ws(client: AuthedClient, ws: WebSocketUpgrade) -> impl IntoResponse {
```

Comment:
I no very little about web sockets, is it part of HTTP or considered a separate protocol? If the latter, then we may not want to include it in the `http` module.

Reply:
Kinda both. I starts the connection as normal HTTP, then requests an upgrade to websockt. I think it belongs here, especially because http and ws are sharing the statement execution and transaction logic.